### PR TITLE
Add option to create amazon ec2 instances with the EbsOptimized setti…

### LINKED ISSUE #1904: ec2 driver should allow to pass flag to request EBSOptimized instance
The default ami used has an EBS file system.  When creating a instance AWS allows you to specify 

"EbsOptimized": [
        "true"
    ],

This results in a much better performing instance as I/O between the host and the disk is much faster.

### PR DESCRIPTION
**Pull request by jeffellin**
fixes #1904 Non EbsOptimized instances will be created by default but adding the flag
UseEbsOptimizedInstance will create and EBSOptimized Instance.

Signed-off-by: jeffellin <jeff@ellin.com>

### CODE REVIEW

**Comment by dmp42** [comment]
@jeffellin You need to fix the formatting of your submission (use `gofmt -w -s` on the problematic files).

**Comment by jeffellin** [comment]
I have a 2nd commit that I already pushed with the requested formatting
fixes.  Should I add that to the PR before doing these steps?  I attempted
to squash the commit on that one but I am a little unsure if that worked or
not.

Thanks

**Comment by jeffellin** [comment]
@dmp42  Been trying to follow along on the various requests here to fix the commit,   I have ammeded the comments, but I am noticed that there is a commit from you as part of the PR still.

**Comment by dmp42** [comment]
@jeffellin you probably need to rebase against master (there is no reason for my commits to be there).

To squash your commits, here is some information: https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

**Comment by jeffellin** [comment]
@dmp42  I think I finally got everything squashed on my end,  but I think I somehow included one of your commits which is now in this PR,   I am not sure how to fix short of closing this PR and starting over.

**Comment by jeffellin** [comment]
I think I have everything squared away now,  just one commit with 3 changed files.  Let me know if there is anything else I need to do.

**Comment by dmp42** [comment]
LGTM

cc @nathanleclaire @ehazlett

**Comment by dmp42** [comment]
@jeffellin can you add tests for this? integration tests are probably the right place for this. I understand it's messy in there, but we really need more "real-life" coverage for this kind of functionality.

**Comment by jeffellin** [comment]
@dmp42 I will take a look,   Also,  I'd like to chat with someone regarding tests for a PR I am creating for another issue I am working on.  Is IRC the best place for that?

**Comment by tamrsolutions** [comment]
@dmp42 Took a look at the integration tests,  doesn't look too bad,  but in order to do a test, you would need an Amazon IAM credential.  Not sure how to handle that. I can certainly work on developing the test but I can't commit it with my credential.

**Comment by jeffellin** [comment]
@dmp42 Took a look at the integration tests, doesn't look too bad, but in order to do a test, you would need an Amazon IAM credential. Not sure how to handle that. I can certainly work on developing the test but I can't commit it with my credential.

**Comment by nathanleclaire** [comment]
@jeffellin You should be able to write the test setting those in environment variables like so:

```
$ DRIVER=virtualbox AWS_ACCESS_KEY_ID=xxx AWS_SECRET_ACCESS_KEY=yyy ./run-bats.sh drivers/amazonec2
```

On our end, we have our own credentials that we'll set in the same fashion.

Note that you will have to add the `amazonec2` directory.  Take a look at the `test/integration/drivers/virtualbox` directory for some examples of driver-specific tests as well.

**Comment by nathanleclaire** [comment]
BTW, I'm usually on IRC during the day (PST), so feel free to ping me (`nathanleclaire`) if you want to chat about testing in real-time.

**Comment by tamrsolutions** [comment]
@nathanleclaire any suggestions on how to assert that the test was successful?  I could use the machine inspect command but that really just tells me the option that was passed?  In order to verify that an EBS instance was actually created I would probably need call an AWS api?  Maybe using curl?

**Comment by jeffellin** [comment]
@nathanleclaire sorry posted that last message from the wrong account.

**Comment by nathanleclaire** [comment]
@jeffellin No prob --  Ideally yes, but curl is tricky since making a "bare" request to the AWS API is complex and involves a variety of signing mechanisms (or AWS tools).  Maybe we can either:

1. Make a call to an instance introspection endpoint run by Amazon over `docker-machine ssh`.  I thought AWS offered something like this, but I'm having trouble digging it up at the moment.
2. Assume that the aws-cli is installed where the integration tests are being run.

**Comment by jeffellin** [comment]
I think I might have found a solution. 

@nathanleclaire 
sudo docker run busybox wget -q -O - "$@"  http://169.254.169.254/latest/meta-data/

see http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html#instancedata-data-retrieva

Now to see if the information I need is actually in the instance metadata

**Comment by jeffellin** [comment]
So, that was a no go,  the info I need isn't there.  It did lead me down another path though.  There is aws docker container with the cli.  I can use that to run the aws cli.

**Comment by nathanleclaire** [comment]
> There is aws docker container with the cli. I can use that to run the aws cli.

Hm, that's an interesting idea.

**Comment by jeffellin** [comment]
Well,  I happen to conveniently have a docker host up and running that I
can run commands on.

**Comment by jeffellin** [comment]
@nathanleclaire I added an integration test.  Not sure if the ci env will run it correctly as there is a bunch of required environment variables for AWS.  Take a look and let me know your thoughts.

File: test/integration/drivers/amazonec2/create-ebsinstance.bats [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+require_env AWS_ACCESS_KEY_ID
+
+require_env AWS_SECRET_ACCESS_KEY
+
+
+@test "$DRIVER: Should Create an EBS Optimized Instance" {
+  
+  run machine --debug create -d amazonec2 --amazonec2-use-ebs-optimized-instance $NAME
+
+  run docker $(machine config $NAME) run --rm -e AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID -e AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY -e AWS_DEFAULT_REGION=$AWS_DEFAULT_REGION blendle/aws-cli ec2 describe-instances --filters Name=tag:Name,Values=$NAME Name=instance-state-name,Values=running --query 'Reservations[0].Instances[0].EbsOptimized' --output text
```
Comment (nathanleclaire) [inline test/integration/drivers/amazonec2/create-ebsinstance.bats]: FYI: You should be able to just do `-e AWS_ACCESS_KEY_ID -e AWS_SECRET_ACCESS_KEY` and achieve the same result (Docker will pass through the corresponding environment variable values).

File: test/integration/drivers/amazonec2/create-ebsinstance.bats [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bats
+
+load ${BASE_TEST_DIR}/helpers.bash
+
+force_env DRIVER amazonec2
+
+#Use Instance Type that supports EBS Optimize 
+export AWS_INSTANCE_TYPE=m4.large
+
+require_env AWS_DEFAULT_REGION
```
Comment (nathanleclaire) [inline test/integration/drivers/amazonec2/create-ebsinstance.bats]: Why is this required?  It should just default to `us-east-1`, right?

**Comment by nathanleclaire** [comment]
Looking pretty good @jeffellin -- nice job.  Don't worry about the CI, we will be implementing that later with our own credentials.

My only concern is the image used --- @dmp42 what do you think about using an image we control like https://hub.docker.com/r/nathanleclaire/aws-cli (a clone of the referenced image) instead?  Then we could use it in the future to when we need `aws-cli` etc. as well.

File: test/integration/drivers/amazonec2/create-ebsinstance.bats (hunk 2 above)
Comment (jeffellin) [inline test/integration/drivers/amazonec2/create-ebsinstance.bats]: It is required by the aws cli.  Otherwise it will complain there is no default region.

**Comment by jeanlaurent** [comment]
@jeffellin The aws driver has been significantly changed by #2684, would you mind rebasing your work on this ?

**Comment by jeffellin** [comment]
This has been sitting a long time. If I do the rebase will it be merged?

**Comment by nathanleclaire** [comment]
@jeffellin I'm in favor of doing so.

**Comment by jeffellin** [comment]
@nathanleclaire what do you want to do with the docker image for the integration test?

**Comment by jeffellin** [comment]
Any github gurus here,  I seem to have gotten all the commits for every rebase operation.  Not sure how to fix that.  The change is relatively small, might be easier to close this PR and standup a new one.

**Comment by jeffellin** [comment]
Also no clue whats going on here.

cannot find package "github.com/smartystreets/go-aws-auth"

that library does not seem to be vendored.

**Comment by jeffellin** [comment]
Never mind  Looks the rebase has seriously gone wrong.

Gotta figure out how to undo it all and start over.   Still might be less work to close this PR and start over.
